Sprint_4 final project

Markq2 opened the pull request:
Выполнено задание для проекта 4го спринта

Markq2 committed:
Sprint_4 final project

kadirovgm commented on .gitignore:
```diff
@@ -0,0 +1,56 @@
+### Python template
+# Byte-compiled / optimized / DLL files
+__pycache__/
```
Можно лучше: в файле гитигнора лучше оставлять только те локальные файлы, что не должны быть запушены в общий репозиторий, в твоем случае ты использовала общий шаблон, это не совсем верно, мы должны избегать использования ненужного кода в проекте.

Markq2 replied:
Исправлено
убрал лишнее из гитигнора

kadirovgm commented on README.md:
```diff
@@ -0,0 +1,61 @@
+Qa python final sprint. 
```
отлично, можешь активнее использовать markdown

kadirovgm commented on conftest.py:
```diff
@@ -0,0 +1,11 @@
+from selenium import webdriver
+from data.data import MAIN_PAGE_URL
+
+
```
супер!

kadirovgm commented on data/data.py:
```diff
@@ -0,0 +1,23 @@
+    '8answer': 'Да, обязательно. Всем самокатов! И Москве, и Московской области.'
+}
+
+
```
отлично! тестовые статичные данные лучше выносить в отдельный файл типа [data.py](http://data.py/) или [urls.py](http://urls.py/)
Это нужно для легкой поддержке проекта и переиспользовании тестовых данных, мы должны стремиться к тому, чтобы разделять код и тестовые данные, тем самым избегая повторений

Markq2 replied:
Согласен, что при большом объеме данных и их стоит разделять по файлам. В данной реализации оставил всё в data

kadirovgm commented on locators/locators.py:
```diff
@@ -0,0 +1,35 @@
+class MainPageLocators:
```
хорошие локаторы!

kadirovgm commented on pages/base_page.py:
```diff
@@ -0,0 +1,54 @@
+from data.data import DZEN_URL, ORDER_PAGE_URL, MAIN_PAGE_URL
+
+
+class BasePage:
```
Отлично! кажется ты правильно понял консепцию POM и базового класса BasePage

класс BasePage - это такой родительский класс, в котором содержатся общие методы, такие как: клик по определенному элементу, скролл к определенному элементу, ожидание определенного элемента, изменение табов итд.

То есть эти методы не привязаны к определенной странице, и могут быть переиспользованы в любой POM-странице, стоит сюда добавить именно такие методы.

Именно поэтому у BasePage не могут быть свои локаторы, ведь предполагается, что для методов BasePage подходят локаторы POM-страницы

kadirovgm commented on pages/base_page.py:
```diff
@@ -0,0 +1,54 @@
+    def press_button(self, by, locator, button):
+        self.driver.find_element(by, locator).send_keys(button)
+
+    def wait_for_url(self, url):
```
нужно исправить: так как это методы базового класса, то реализацию специфичных сценариев в классе BasePage уже быть не должно. Здесь нужно сделать общий метод, который принимает на вход URL и сравнивает с тем, что ты задала (тоже как аргумент функции)

т.е. в POM страницах, например, в main_page.py можешь уже сделать метод, который проверяет ссылки, с помощью базового метода wait_for_url (current_url, expected_url)

Markq2 replied:
Исправлено. Удален метод wait_for_url и его использования в тестах на проверку редиректа. 
В текущей реализации не требуется ожидание url, как мне кажется достаточно ожидания элемента на конечной странице и сравнения текущего урла с ожиданием с помощью get_current_url

kadirovgm commented on pages/main_page.py:
```diff
@@ -0,0 +1,36 @@
+import allure
+
+
+class MainPage(BasePage):
```
молодец, что наследуешь

kadirovgm commented on pages/main_page.py:
```diff
@@ -0,0 +1,36 @@
+
+class MainPage(BasePage):
+    @allure.step('Скролл страницы до вопросов')
+    def scroll_page_down(self):
```
нужно исправить, так как ты скроллишь к определенному элементу, то нужно в названии метода это явно указать, иначе можно сделать общий метод в класс BasePage, который будет уметь скроллить в любой нужный элемент, который задать в аргументах функции

Markq2 replied:
Название исправлено на scroll_page_down_to_questions_list

kadirovgm commented on pages/main_page.py:
```diff
@@ -0,0 +1,36 @@
+        BasePage.scroll_to_element(self, By.CSS_SELECTOR, MainPageLocators.questions)
+
+    @allure.step('Клик по вопросу {question}')
+    def question_click(self, question):
```
отлично!

kadirovgm commented on pages/order_page.py:
```diff
@@ -0,0 +1,107 @@
+        BasePage.click(self, By.XPATH, OrderPageLocators.create_order_button)
+
+    @allure.step('Открытие формы заказа по кнопке, заполнение данных заказа')
+    def create_order(self, button, color):
```
супер, молодец, что объединил эти шаги!

kadirovgm commented on tests/test_main_page.py:
```diff
@@ -0,0 +1,28 @@
+        main_page = MainPage(driver)
+        main_page.scroll_page_down()
+        main_page.question_click(question)
+        assert answer == main_page.get_answer_text(question), "Ответ не соответствует ожидаемому"
```
хороший тест!

kadirovgm commented:
Кстати, добавь аллюр отчет к своему проекту - забыл важное замечание сказать

Markq2 committed:
Исправление замечаний
Добавлены json'ы для аллюр-отчета
Изменено название метода для скролла к низу страницы
Удален метод ожидания url

Markq2 commented:
> Кстати, добавь аллюр отчет к своему проекту - забыл важное замечание сказать

Проглядел, спасибо. Добавлены json'ы

kadirovgm reviewed (approved):
Молодец!
